src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -1438,185 +1449,169 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         place: &Place<'tcx>,
         including_downcast: IncludingDowncast,
     ) -> Option<String> {
-        let mut buf = String::new();
-        match self.append_place_to_string(place, &mut buf, false, &including_downcast) {
-            Ok(()) => Some(buf),
-            Err(()) => None,
-        }
+        let neo_place = self.infcx.tcx.as_new_place(place);
+        self.append_place_to_string(&neo_place, &including_downcast).ok()
     }
 
     /// Appends end-user visible description of `place` to `buf`.
     fn append_place_to_string(
         &self,
-        place: &Place<'tcx>,
-        buf: &mut String,
-        mut autoderef: bool,
+        place: &NeoPlace<'tcx>,
         including_downcast: &IncludingDowncast,
-    ) -> Result<(), ()> {
-        match *place {
-            Place::Promoted(_) => {
-                buf.push_str("promoted");
-            }
-            Place::Local(local) => {
-                self.append_local_to_string(local, buf)?;
-            }
-            Place::Static(ref static_) => {
-                buf.push_str(&self.infcx.tcx.item_name(static_.def_id).to_string());
+    ) -> Result<String, ()> {
+        debug!("describe_place: {:?}", place);
+
+        let mut place_str = String::new();
+        debug!("describe_place-base: {:?}", &place.base);
+        let no_local_name = if self.mir.local_decls.is_empty() {
+            true
+        } else {
+            if self.mir.local_decls.len() > 1 {
+                self.mir.local_decls[Local::new(1)].name.is_none()
+            } else {
+                false
             }
-            Place::Projection(ref proj) => {
-                match proj.elem {
-                    ProjectionElem::Deref => {
-                        let neo_place = self.infcx.tcx.as_new_place(place);
-                        let upvar_field_projection =
-                            neo_place.is_upvar_field_projection(self.mir, &self.infcx.tcx);
-                        if let Some(field) = upvar_field_projection {
-                            let var_index = field.index();
-                            let name = self.mir.upvar_decls[var_index].debug_name.to_string();
-                            if self.mir.upvar_decls[var_index].by_ref {
-                                buf.push_str(&name);
-                            } else {
-                                buf.push_str(&format!("*{}", &name));
+        };
+        if PlaceBase::Local(Local::new(1)) == place.base && no_local_name {
```

Comment:
In particular i'm not sure what this special case corresponds to -- we don't seem to have such a thing before?

Reply:
it's kind of tricky, I'll explain in zulip.